### ISSUE TITLE
Pretraining BART language model

### ISSUE THREAD

**Issue by sajastu** [huggingface/transformers#18030]
### Feature request

Hi,

I'm looking into BART [docs](https://huggingface.co/docs/transformers/model_doc/bart). Seems that the provided examples are on fine-tuning BART on Seq2Seq summarization tasks. I'm wondering if there is any example on pertaining BART's "Language Model" itself, with the pre-training objectives (Token infilling, Token Masking, etc.) that are mentioned in the original paper. I was looking into this a couple of months ago and found this thread: https://github.com/huggingface/transformers/issues/4151 and a relevant issue in fairseq: https://github.com/facebookresearch/fairseq/issues/1899. Now decided to ask it directly here to see if there has been any update so far. 

Thanks, @patrickvonplaten @patil-suraj,

### Motivation

Making BART for further pre-training (on Language Model).

**Comment by Lauler** [comment]
I have attempted to reproduce BART pretraining for Swedish using fairseq [here](https://github.com/kb-labb/kb_bart) with some instructions in the README. We trained a custom tokenizer with Huggingface `tokenizers` package and figured we'd later retrofit the `dict.txt` generated by fairseq preprocessing back to a Huggingface compatible `tokenizer.json`. This worked, but we later learned it was unnecessarily complicated, as we could have just copy pasted the vocabulary from our original `tokenizer.json` to a tab separated `dict.txt` where the 1st column is the tokens, and the 2nd column can be dummy frequency counts of how often the tokens occur in your vocab (You can set the frequencies to 0 or any integer). 

We were unfamiliar with fairseq prior to doing this and based the entire reproduction off of reading the paper closely and trying to match the reported experimental settings in the paper against fairseq args by reading the documentation and the source code.

The finished model was uploaded to [Huggingface](https://huggingface.co/KBLab/bart-base-swedish-cased). However, we learned quite a few lessons along the way that aren't documented there. 

The first one being the easiest process of using a custom Huggingface tokenizer with fairseq that I described above.

A second important one was that you should generate several epochs worth of sharded and shuffled data and not just a shards for one single epoch. Fairseq pretraining will crash once you reach the end of your listed shards, and won't allow you to reuse or relist the same shard filename. I guess this sort of stuff maybe is obvious if you have the chance to ask your Facebook engineer colleagues questions over a lunch, but to us it wasn't obvious what best practices were, since it's not described in the docs. 

Anyway. Maybe the pretraining task along with the training args I researched in trying to replicate it may be of interest to you: https://github.com/kb-labb/kb_bart/blob/main/train_bart_args.sh . I cannot guarantee that it is identical to actual BART pretraining settings, but it was my best attempt based on reading the paper and source code.

**Comment by ccdv-ai** [comment]
@sajastu 
You can use this [repo](https://github.com/morganmcg1/rotobart/blob/main/data_collator.py) and in the HF [run_mlm.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_mlm.py) script you have to:
* replace `AutoModelForMaskedLM` by `AutoModelForSeq2SeqLM` 
* replace `data_collator` by `DataCollatorForDenoisingTasks` (token infilling + sentence permutation) 
* or replace `data_collator` by `DataCollatorForTextInfilling` (you need to process decoder_input_ids for this one)  

I'm not the author of this repo but @patrickvonplaten is one of them. 
You need to convert numpy outputs from the collator to torch ones if you dont want to rewrite it.
You can also check [#12370](https://github.com/huggingface/transformers/pull/12370) for the pytorch implementation of the `DataCollatorForTextInfilling` which is very similar.

 That's it!

**Comment by PiotrNawrot** [comment]
We've released [nanoT5](https://github.com/PiotrNawrot/nanoT5) that reproduces T5-model (similar to BART) pre-training. You can take a look! Any suggestions are more than welcome.

**Comment by ledaat** [comment]
A simple question: when pre-training (denoising) with mBART-50, should decoder_input_ids be based upon **labels** (the correct text) or upon **input_ids** (the corrupted text)?

In the code by @patrickvonplaten that @ccdv-ai referenced above, decoder_input_ids appears to be input_ids shifted right:       [batch["decoder_input_ids"] = self.shift_tokens_right(batch["input_ids"])](https://github.com/morganmcg1/rotobart/blob/cd29e458d0adfa270d47e11614427935541aef0c/data_collator.py#L247).

However, from many other sources, it seems that decoder_input_ids is always based on labels, from which the model learns via cross-attention. (See also https://github.com/huggingface/transformers/issues/11357 where it is claimed that the mention of input_ids is a bug, as labels is actually used).

So which one is it? I would be immensely grateful for any clarification or guidance!
